### PR TITLE
Server/Channel Improvements

### PR DESCRIPTION
**Pull request by SnazzyPine25**
This Pull Request Includes
- Verification Levels
- Server Integrations
- Widget URL / Widget Banner URL
- Splash URL
- Icon/Avatar Checking

### CODE REVIEW

File: lib/discordrb/api.rb [hunk 1]
```diff
@@ -103,6 +103,16 @@ def icon_url(server_id, icon_id)
     "#{api_base}/guilds/#{server_id}/icons/#{icon_id}.jpg"
   end
 
+  # Make an widget URL from server IDs
```
Comment (meew0) [inline lib/discordrb/api.rb]: "a widget URL", and it's only one ID

File: lib/discordrb/api.rb [hunk 2]
```diff
@@ -103,6 +103,16 @@ def icon_url(server_id, icon_id)
...
+  def widget_url(server_id)
+    "#{api_base}/guilds/#{server_id}/widget.png"
+  end
+
+  # Make an widget URL from server IDs
```
Comment (meew0) [inline lib/discordrb/api.rb]: "a widget URL"

File: lib/discordrb/data.rb [hunk 3]
```diff
@@ -903,9 +904,19 @@ class Channel
     # @return [Recipient, nil] the recipient of the private messages, or nil if this is not a PM channel
     attr_reader :recipient
 
-    # @return [String] the channel's topic
+    # @return [String, nil] the channel's topic. `nil` if channel is voice
     attr_reader :topic
 
+    # @return [Message, nil] the last message sent to this channel. `nil` if channel is voice or no last message found
+    attr_reader :last_message
```
Comment (meew0) [inline lib/discordrb/data.rb]: Why is this necessary? This will only be set once at creation and be useless afterwards.

File: lib/discordrb/data.rb [hunk 4]
```diff
@@ -1024,10 +1038,28 @@ def name=(name)
...
     def topic=(topic)
+      raise ArgumentError, 'This channel is not a text channel!' unless text?
       @topic = topic
       update_channel_data
     end
 
+    # Sets this channel's bitrate.
+    # @param bitrate [Integer] The new bitrate. Number has to be between 8-96
+    def bitrate=(bitrate)
+      raise ArgumentError, 'This channel is not a voice channel!' unless voice?
```
Comment (meew0) [inline lib/discordrb/data.rb]: This shouldn't be an ArgumentError as it's not a problem with the arguments; you can just pass on Discord's API error here.

File: lib/discordrb/data.rb [hunk 5]
```diff
@@ -1024,10 +1038,28 @@ def name=(name)
     # Sets this channel's topic.
     # @param topic [String] The new topic.
     def topic=(topic)
+      raise ArgumentError, 'This channel is not a text channel!' unless text?
```
Comment (meew0) [inline lib/discordrb/data.rb]: See https://github.com/meew0/discordrb/pull/181/files#r74666895

File: lib/discordrb/data.rb [hunk 6]
```diff
@@ -1024,10 +1038,28 @@ def name=(name)
...
+    def bitrate=(bitrate)
+      raise ArgumentError, 'This channel is not a voice channel!' unless voice?
+      @bitrate = bitrate
+      update_channel_data
+    end
+
+    # Sets this channel's user limit.
+    # @param limit [Integer] The new user limit. `0` for unlimited, Has to be a number between 0-99
+    def user_limit=(limit)
+      raise ArgumentError, 'This channel is not a voice channel!' unless voice?
```
Comment (meew0) [inline lib/discordrb/data.rb]: See https://github.com/meew0/discordrb/pull/181/files#r74666895

File: lib/discordrb/data.rb [hunk 7]
```diff
@@ -1024,10 +1038,28 @@ def name=(name)
...
+    # Sets this channel's bitrate.
+    # @param bitrate [Integer] The new bitrate. Number has to be between 8-96
+    def bitrate=(bitrate)
+      raise ArgumentError, 'This channel is not a voice channel!' unless voice?
+      @bitrate = bitrate
+      update_channel_data
+    end
+
+    # Sets this channel's user limit.
+    # @param limit [Integer] The new user limit. `0` for unlimited, Has to be a number between 0-99
```
Comment (meew0) [inline lib/discordrb/data.rb]: "has" should not be capitalised

File: lib/discordrb/data.rb [hunk 8]
```diff
@@ -1344,6 +1376,7 @@ module ServerAttributes
     # Utility function to get the URL for the icon image
     # @return [String] the URL to the icon image
     def icon_url
+      return nil unless @icon_id
```
Comment (meew0) [inline lib/discordrb/data.rb]: Why is this check here? As far as I know the icon ID will never be nil.

File: lib/discordrb/data.rb [hunk 9]
```diff
@@ -1365,11 +1398,18 @@ class Server
     # @return [Array<Role>] an array of all the roles created on this server.
     attr_reader :roles
 
+    # @return [String] the hexadecimal ID used to identify this server's splash.
```
Comment (meew0) [inline lib/discordrb/data.rb]: "splash" should be clarified

File: lib/discordrb/data.rb [hunk 10]
```diff
@@ -1468,6 +1510,18 @@ def voice_channels
       @channels.select(&:voice?)
     end
 
+    # @return [String, nil] The widget URL to the server. `nil` if it's not embed.
```
Comment (meew0) [inline lib/discordrb/data.rb]: Don't capitalise the text for `@return` tags, otherwise it looks weird in Rubydoc ("Returns The widget URL ..."). It should also ideally be one sentence, how about "the server's widget URL, or nil if the server doesn't have the widget enabled"?

File: lib/discordrb/data.rb [hunk 11]
```diff
@@ -903,9 +904,19 @@ class Channel
...
+    # @return [String, nil] the channel's topic. `nil` if channel is voice
     attr_reader :topic
 
+    # @return [Message, nil] the last message sent to this channel. `nil` if channel is voice or no last message found
+    attr_reader :last_message
+
+    # @return [Integer, nil] the bitrate of the channel. `nil` if channel is text
+    attr_reader :bitrate
+
+    # @return [Integer, nil] the amount users that can be in the channel. `0` means it is unlimited. `nil` if channel is text
```
Comment (meew0) [inline lib/discordrb/data.rb]: "the amount *of* users"

File: lib/discordrb/data.rb [hunk 12]
```diff
@@ -1028,6 +1038,21 @@ def topic=(topic)
       update_channel_data
     end
 
+    # Sets this channel's bitrate.
+    # @param bitrate [Integer] The new bitrate. Number has to be between 8-96
+    def bitrate=(bitrate)
+      @bitrate = bitrate
+      update_channel_data
+    end
```
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: Maybe don't just return nil if called on a text channel

File: lib/discordrb/data.rb [hunk 13]
```diff
@@ -1028,6 +1038,21 @@ def topic=(topic)
       update_channel_data
     end
 
+    # Sets this channel's bitrate.
+    # @param bitrate [Integer] The new bitrate. Number has to be between 8-96
```
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: Bitrate has to be between 8000 and 96000 (you missed a few zeros :smile: )

File: lib/discordrb/data.rb [hunk 14]
```diff
@@ -1028,6 +1038,21 @@ def topic=(topic)
...
+      update_channel_data
+    end
+
+    # Sets this channel's user limit.
+    # @param limit [Integer] The new user limit. `0` for unlimited, has to be a number between 0-99
+    def user_limit=(limit)
+      @user_limit = limit
+      update_channel_data
+    end
+    alias_method :limit=, :user_limit=
```
Comment (greenbigfrog) [inline lib/discordrb/data.rb]:
```
[19] pry(main)> bot.channel(215741090228207619).user_limit=5
RestClient::BadRequest: 400 Bad Request
from /home/jonathan/.rvm/gems/ruby-head/gems/rest-client-2.0.0/lib/restclient/abstract_response.rb:223:in `exception_with_response'
```
bot got administrator permission, so it has to do something with the call...

File: lib/discordrb/data.rb [hunk 15]
```diff
@@ -1445,6 +1544,11 @@ def members
       @members.values
     end
 
+    # @return [Array<Integration>] an array of all the intergrations connected to this server.
+    def integrations
+      API.server_integration(@bot.token, @id).map { |element| Integration.new(element) }
+    end
+
```
Comment (greenbigfrog) [inline lib/discordrb/data.rb]:
```
[22] pry(main)> bot.server(215741090228207618).integrations
NoMethodError: undefined method `server_integration' for Discordrb::API:Module
from /home/jonathan/.rvm/gems/ruby-head/gems/discordrb-2.1.3/lib/discordrb/data.rb:1549:in `integrations'
```

File: lib/discordrb/data.rb [hunk 16]
```diff
@@ -1468,6 +1572,28 @@ def voice_channels
       @channels.select(&:voice?)
     end
 
+    # @return [String, nil] the widget URL to the server that displays the amount of online members in a
+    # stylish way. `nil` if it's not embed.
+    def widget_url
+      return nil unless @embed
+      API.widget_url(@id)
+    end
```
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: this always returns nil for me?

Comment (greenbigfrog) [inline lib/discordrb/data.rb]: https://cdn.discordapp.com/attachments/83281822225530880/216211893289746432/IMG_2723.PNG (copyed for snazzah)

30 minutes later `bot.server(215741090228207618).widget_url` still returns `nil`

Comment (greenbigfrog) [inline lib/discordrb/data.rb]: issue is probably that `@embed` is always false/nil and won't turn true

Comment (greenbigfrog) [inline lib/discordrb/data.rb]: https://discordapp.com/widget?id=215741090228207618&theme=dark
(If you wanna check yourself that widget loads perfectly)

File: lib/discordrb/data.rb [hunk 17]
```diff
@@ -1028,6 +1038,21 @@ def topic=(topic)
       update_channel_data
     end
 
+    # Sets this channel's bitrate.
+    # @param bitrate [Integer] The new bitrate. Number has to be between 8000-96000 (128000 for VIP servers)
+    def bitrate=(bitrate)
+      @bitrate = bitrate
+      update_channel_data
+    end
```
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: Hence my commit, my old comment got marked as on outdated diff.

Still not sure if this should return sth else then nil for text channels @meewo

File: lib/discordrb/data.rb (hunk 8 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]:
```
bot.profile.avatar_url
=> "https://discordapp.com/api/users/215739751611891715/avatars/.jpg"
```
This isn't a picture or is it?

File: lib/discordrb/data.rb (hunk 14 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: Working now... No idea why...

Also the question here if it should return sth diffrent then `nil` for textchannels @meew0

File: lib/discordrb/data.rb (hunk 8 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]:
```
bot.server(215741090228207618).icon_url
=> "https://discordapp.com/api/guilds/215741090228207618/icons/.jpg"
```
This isn't a valid pic...

File: lib/discordrb/data.rb [hunk 18]
```diff
@@ -1365,14 +1451,27 @@ class Server
...
+    attr_reader :splash_id
+
     # @return [true, false] whether or not this server is large (members > 100). If it is,
     # it means the members list may be inaccurate for a couple seconds after starting up the bot.
     attr_reader :large
     alias_method :large?, :large
 
+    # @return [true, false] whether or not the server has widget enabled
+    attr_reader :embed
+    alias_method :embed?, :embed
```
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: always returns `nil` for me

File: lib/discordrb/data.rb [hunk 19]
```diff
@@ -1365,14 +1451,27 @@ class Server
...
+
+    # @return [Array<String>] the features of the server
+    attr_reader :features
```
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: What could be returned here @SnazzyPine25 ?

File: lib/discordrb/data.rb [hunk 20]
```diff
@@ -1445,6 +1548,11 @@ def members
       @members.values
     end
 
+    # @return [Array<Integration>] an array of all the intergrations connected to this server.
+    def integrations
+      API.server_integrations(@bot.token, @id).map { |element| Integration.new(element) }
+    end
```
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: bot.server(215741090228207618).integrations
NoMethodError: undefined method `map' for <RestClient::Response 200 "[]">:RestClient::Response
from /home/jonathan/.rvm/gems/ruby-head/gems/discordrb-2.1.3/lib/discordrb/data.rb:1553:in `integrations'

File: lib/discordrb/data.rb (hunk 19 above)
Comment (SnazzyPine25) [inline lib/discordrb/data.rb]: `array of guild features`

File: lib/discordrb/data.rb [hunk 21]
```diff
@@ -1391,6 +1490,10 @@ def initialize(data, bot, exists = true)
 
       @large = data['large']
       @member_count = data['member_count']
+      @verification_level = [:none, :low, :medium, :high][data['verification_level']]
```
Comment (SnazzyPine25) [inline lib/discordrb/data.rb]: The array is zero-indexed, this would be displaced

Comment (greenbigfrog) [inline lib/discordrb/data.rb]: Yes... first entry of array is at [0]

File: lib/discordrb/data.rb (hunk 8 above)
Comment (meew0) [inline lib/discordrb/data.rb]: Of course..

File: lib/discordrb/data.rb (hunk 14 above)
Comment (meew0) [inline lib/discordrb/data.rb]: Ideally throw some descriptive exception, maybe make a new one if it's just a 400

File: lib/discordrb/data.rb (hunk 17 above)
Comment (meew0) [inline lib/discordrb/data.rb]: (See above)

File: lib/discordrb/data.rb (hunk 19 above)
Comment (meew0) [inline lib/discordrb/data.rb]: Make this symbols instead of strings, please, as it will be a set of fixed values.

File: lib/discordrb/data.rb (hunk 20 above)
Comment (meew0) [inline lib/discordrb/data.rb]: This needs to be parsed, obviously

**Comment by greenbigfrog** [comment]
Hence discords bad backend programming the servers don't error if you try to set a topic on a voice channel for example. Not sure how I should implement differing of voice and text channels (or throwing errors in this case).

**Comment by meew0** [comment]
You can just raise a `RuntimeError` with a message that you can't set the topic on a voice channel, in that case.

**Comment by PoVa** [comment]
Wouldn't it be better to make `Channel` class that would have stuff that's usable for both voice and text channels and then create `VoiceChannel` and `TextChannel` classes that have methods unique to them (for example `VoiceChannel#bitrate=`)

**Comment by greenbigfrog** [comment]
That was what I also had thought about....

**Comment by meew0** [comment]
That could be done but it would be awkward to implement because you would have to duplicate stuff like the `API` method. You would also be fighting against the API (for which both classes of channels are the same) which is something I try to avoid, since it leads to susceptibility to breakage in the case of slight changes to the API.

**Comment by greenbigfrog** [comment]
Ok. So keep it as it is right now!?

File: lib/discordrb/data.rb (hunk 21 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: (This is fixed in ae1a595 )

**Comment by meew0** [comment]
Yes. I'll merge it as soon as the outstanding problems are fixed

**Comment by greenbigfrog** [comment]
Creating integrations (f.e.) isn't implemented yet... https://discordapp.com/developers/docs/resources/guild#create-guild-integration

(maybe we should delay that in a other PR)

**Comment by meew0** [comment]
Well, none of the integration editing stuff (create/update/delete) is implemented - I'm fine with it being in a later PR.

File: lib/discordrb/data.rb [hunk 22]
```diff
@@ -1365,14 +1452,27 @@ class Server
     # @return [Array<Role>] an array of all the roles created on this server.
     attr_reader :roles
 
+    # @return [String] the hexadecimal ID used to identify this server's splash image for their VIP invite page.
+    attr_reader :splash_id
```
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: haven't been able to test this...

Comment (meew0) [inline lib/discordrb/data.rb]: Try the /r/Overwatch server, I'm pretty sure it has a custom splash.

Comment (greenbigfrog) [inline lib/discordrb/data.rb]: But how do I get my bot there? I tought you don't like user bots :)

File: lib/discordrb/data.rb [hunk 23]
```diff
@@ -1447,6 +1547,20 @@ def members
...
+    # @return [Array<Integration>, nil] an array of all the intergrations connected to this server (nil if there's no integration).
+    def integrations
+      integration = JSON.parse(API.server_integrations(@bot.token, @id))
+      return nil if integration.empty?
+      integration.map { |element| Integration.new(element) }
+    end
+
+    # @return [true, false] whether or not the server has widget enabled
+    def embed
+      @embed = JSON.parse(API.server(@bot.token, @id))['embed_enabled'] if @embed.nil?
```
Comment (meew0) [inline lib/discordrb/data.rb]: You still need to return `@embed` here, otherwise it will just do the resolution and not return the value. Also, now that it's not an `attr_reader` anymore, you can get rid of the alias and just rename this method to `embed?`. The actual caching is fine.

File: lib/discordrb/data.rb (hunk 16 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: Fixed, since I fixed @embed always being nil

**Comment by greenbigfrog** [comment]
The Integration part hasn't been tested yet. (In theory it should work, but testing is better)

Else the Integration part could be moved to an extra PR in total...

**Comment by meew0** [comment]
Honestly, it doesn't matter that much - the code looks correct to me, and as long as it doesn't spectacularly break everything else, I'll just merge it. It's unlikely that this code will get used much, and people who do use it and experience bugs can just report them.

**Comment by meew0** [comment]
I'll just go through the comments and point out typos, then it should be ready to merge

File: lib/discordrb/api.rb [hunk 24]
```diff
@@ -103,6 +103,16 @@ def icon_url(server_id, icon_id)
     "#{api_base}/guilds/#{server_id}/icons/#{icon_id}.jpg"
   end
 
+  # Make an widget picture URL from server ID
```
Comment (meew0) [inline lib/discordrb/api.rb]: "a widget picture"

File: lib/discordrb/api.rb [hunk 25]
```diff
@@ -103,6 +103,16 @@ def icon_url(server_id, icon_id)
...
+  def widget_url(server_id)
+    "#{api_base}/guilds/#{server_id}/widget.png"
+  end
+
+  # Make an splash URL from server and splash IDs
```
Comment (meew0) [inline lib/discordrb/api.rb]: "a splash URL"

**Comment by greenbigfrog** [comment]
Ok. I personally prefer testing first, but it's not me who's responsible for this lib `¯\_(ツ)_/¯`

File: lib/discordrb/data.rb [hunk 26]
```diff
@@ -906,6 +904,13 @@ class Channel
     # @return [String] the channel's topic
     attr_reader :topic
 
+    # @return [Integer] the bitrate of the channel
```
Comment (meew0) [inline lib/discordrb/data.rb]: This comment should specify the unit the bitrate is in

File: lib/discordrb/data.rb [hunk 27]
```diff
@@ -1024,10 +1031,29 @@ def name=(name)
     # Sets this channel's topic.
     # @param topic [String] The new topic.
     def topic=(topic)
+      raise "Voice channels don't have topics" if voice?
       @topic = topic
       update_channel_data
     end
 
+    # Sets this channel's bitrate.
+    # @param bitrate [Integer] The new bitrate. Number has to be between 8000-96000 (128000 for VIP servers)
```
Comment (meew0) [inline lib/discordrb/data.rb]: This one should also specify the unit

File: lib/discordrb/data.rb [hunk 28]
```diff
@@ -1024,10 +1031,29 @@ def name=(name)
     # Sets this channel's topic.
     # @param topic [String] The new topic.
     def topic=(topic)
+      raise "Voice channels don't have topics" if voice?
```
Comment (meew0) [inline lib/discordrb/data.rb]: I'd prefer if this were in line with the other comments, i.e. `'Tried to send topic on voice channel'`

File: lib/discordrb/data.rb [hunk 29]
```diff
@@ -1344,10 +1370,76 @@ module ServerAttributes
...
+    # @return [Server] the server the integration is linked to
+    attr_reader :server
+
+    # @return [User] the user the integration is linked to
+    attr_reader :user
+
+    # @return [Role, nil] the role that this integration uses for "subscribers"
+    attr_reader :role
+
+    # @return [true, false] wheter emoticons are enabled
```
Comment (meew0) [inline lib/discordrb/data.rb]: "whether"

File: lib/discordrb/data.rb [hunk 30]
```diff
@@ -1344,10 +1370,76 @@ module ServerAttributes
...
+    attr_reader :emoticon
+    alias_method :emoticon?, :emoticon
+
+    # @return [String] the integration type `youtube, twitch, etc.`
```
Comment (meew0) [inline lib/discordrb/data.rb]: Wrong use of backticks, this should be "the integration type (YouTube, Twitch, etc.)"

File: lib/discordrb/data.rb [hunk 31]
```diff
@@ -1344,10 +1370,76 @@ module ServerAttributes
...
+    attr_reader :type
+
+    # @return [true, false] whether the integration is enabled
+    attr_reader :enabled
+
+    # @return [true, false] whether the integration is syncing
+    attr_reader :syncing
+
+    # @return [Hash] the integration account information, listing the `name` and `id` in the hash
```
Comment (meew0) [inline lib/discordrb/data.rb]: This should probably be separated into an `IntegrationAccount` class that simply has an ID and name - exposing internal hashes like this is bad style

File: lib/discordrb/data.rb [hunk 32]
```diff
@@ -1344,10 +1370,76 @@ module ServerAttributes
...
+    attr_reader :account
+
+    # @return [Time] the time the integration was synced at
+    attr_reader :synced_at
+
+    # @return [Integer] the behavior of expiring subscribers (0 = Remove User from role; 1 = Kick User from server)
+    attr_reader :expire_behavior
+
+    # @return [Integer] the grace period before expiring subscribers (in days)
```
Comment (meew0) [inline lib/discordrb/data.rb]: "before subscribers expire"

File: lib/discordrb/data.rb [hunk 33]
```diff
@@ -1344,10 +1370,76 @@ module ServerAttributes
...
+    # @return [true, false] whether the integration is syncing
+    attr_reader :syncing
+
+    # @return [Hash] the integration account information, listing the `name` and `id` in the hash
+    attr_reader :account
+
+    # @return [Time] the time the integration was synced at
+    attr_reader :synced_at
+
+    # @return [Integer] the behavior of expiring subscribers (0 = Remove User from role; 1 = Kick User from server)
```
Comment (meew0) [inline lib/discordrb/data.rb]: "behaviour" please :gb:. This should also probably be a symbol instead of just exposing the number, similarly to the verification level.

File: lib/discordrb/data.rb [hunk 34]
```diff
@@ -1344,10 +1370,76 @@ module ServerAttributes
...
+      @expire_behavior = data['expire_behavior']
+      @expire_grace_period = data['expire_grace_period']
+      @user = @bot.ensure_user(data['user'])
+      @role = server.role(data['role_id']) || nil
+      @emoticon = data['enable_emoticons']
+    end
+
+    # The inspect method is overwritten to give more useful output
+    def inspect
+      "<integration name=#{@name} id=#{@id} type=#{@type} enabled=#{@enabled}>"
```
Comment (meew0) [inline lib/discordrb/data.rb]: `Integration` should be capitalised

File: lib/discordrb/data.rb [hunk 35]
```diff
@@ -1370,9 +1462,15 @@ class Server
     attr_reader :large
     alias_method :large?, :large
 
+    # @return [Array<Symbol>] the features of the server
```
Comment (meew0) [inline lib/discordrb/data.rb]: This comment should give some examples of features

File: lib/discordrb/data.rb [hunk 36]
```diff
@@ -1370,9 +1462,15 @@ class Server
...
+    attr_reader :features
+
     # @return [Integer] the absolute number of members on this server, offline or not.
     attr_reader :member_count
 
+    # @return [Symbol] the verification level of the server.
```
Comment (meew0) [inline lib/discordrb/data.rb]: This comment should clarify what exactly the symbols can be

File: lib/discordrb/data.rb [hunk 37]
```diff
@@ -1447,6 +1549,19 @@ def members
 
     alias_method :users, :members
 
+    # @return [Array<Integration>, nil] an array of all the intergrations connected to this server (nil if there's no integration).
```
Comment (meew0) [inline lib/discordrb/data.rb]: "integrations", also, what's wrong with simply returning an empty array in case of no integrations? Simplifies the logic on both discordrb's and the user's side

File: lib/discordrb/data.rb [hunk 38]
```diff
@@ -1468,6 +1583,35 @@ def voice_channels
       @channels.select(&:voice?)
     end
 
+    # @return [String, nil] the widget URL to the server that displays the amount of online members in a
+    # stylish way. `nil` if it's not embed.
```
Comment (meew0) [inline lib/discordrb/data.rb]: "if the widget is not enabled"

File: lib/discordrb/data.rb [hunk 39]
```diff
@@ -1468,6 +1583,35 @@ def voice_channels
...
+    def widget_url
+      return nil unless @embed
+      API.widget_url(@id)
+    end
+
+    # @param style_number [Integer] the style number of the widget picture
+    # @return [String, nil] the widget banner  URL to the server that displays the amount of online members,
+    # server icon and server name in a stylish way. `nil` if it's not embed.
```
Comment (meew0) [inline lib/discordrb/data.rb]: "if the widget is not enabled"

File: lib/discordrb/data.rb [hunk 40]
```diff
@@ -1468,6 +1583,35 @@ def voice_channels
...
 
+    # @return [String, nil] the widget URL to the server that displays the amount of online members in a
+    # stylish way. `nil` if it's not embed.
+    def widget_url
+      return nil unless @embed
+      API.widget_url(@id)
+    end
+
+    # @param style_number [Integer] the style number of the widget picture
+    # @return [String, nil] the widget banner  URL to the server that displays the amount of online members,
```
Comment (meew0) [inline lib/discordrb/data.rb]: There's a double space here

File: lib/discordrb/data.rb [hunk 41]
```diff
@@ -1468,6 +1583,35 @@ def voice_channels
...
     end
 
+    # @return [String, nil] the widget URL to the server that displays the amount of online members in a
+    # stylish way. `nil` if it's not embed.
+    def widget_url
+      return nil unless @embed
+      API.widget_url(@id)
+    end
+
+    # @param style_number [Integer] the style number of the widget picture
```
Comment (meew0) [inline lib/discordrb/data.rb]: There should be information about what style numbers are valid and which number leads to what style

**Comment by meew0** [comment]
Also, yes, of course I'll test it before merging

File: lib/discordrb/data.rb (hunk 31 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: What exactly do you mean?

**Comment by greenbigfrog** [comment]
I should've fixed all issues besides the one, where I need clarification.

File: lib/discordrb/data.rb (hunk 31 above)
Comment (SnazzyPine25) [inline lib/discordrb/data.rb]: @meew0 :ok_hand:

**Comment by meew0** [comment]
@greenbigfrog what is the one where you need clarification? There are no open comments other than the icon ID one (where I was wrong) and the zero-index one (that you claimed was fixed).

**Comment by greenbigfrog** [comment]
Snazzah commited it after I went to bed in efb0f81

So this should be ready...

File: lib/discordrb/data.rb [hunk 42]
```diff
@@ -1447,6 +1564,18 @@ def members
 
     alias_method :users, :members
 
+    # @return [Array<Integration>] an array of all the Intergrations connected to this server.
```
Comment (meew0) [inline lib/discordrb/data.rb]: should be "integrations"

File: lib/discordrb/data.rb [hunk 43]
```diff
@@ -1468,6 +1597,35 @@ def voice_channels
...
+    # @return [String, nil] the widget URL to the server that displays the amount of online members in a
+    # stylish way. `nil` if the widget is not enabled.
+    def widget_url
+      return nil unless @embed
+      API.widget_url(@id)
+    end
+
+    # @param style_number [Integer] the style number of the widget picture  (For the diffrent styles just try which looks how)
+    # @return [String, nil] the widget banner URL to the server that displays the amount of online members,
+    # server icon and server name in a stylish way. `nil` if the widget is not enabled.
```
Comment (meew0) [inline lib/discordrb/data.rb]: The second line here should be indented with two spaces so YARD knows it's still part of the return statement.
```
# @return [String, nil] ...
#   server icon and ...
```

File: lib/discordrb/data.rb [hunk 44]
```diff
@@ -1468,6 +1597,35 @@ def voice_channels
...
+    end
+
+    # @return [String] the hexadecimal ID used to identify this server's splash image for their VIP invite page.
+    def splash_id
+      @splash_id = JSON.parse(API.server(@bot.token, @id))['splash'] if @splash_id.nil?
+      @splash_id
+    end
+
+    # @return [String, nil] the splash image URL for the server's VIP invite page.
+    # `nil` if there is no splash image.
```
Comment (meew0) [inline lib/discordrb/data.rb]: Should be indented too

File: lib/discordrb/data.rb [hunk 45]
```diff
@@ -1468,6 +1597,35 @@ def voice_channels
       @channels.select(&:voice?)
     end
 
+    # @return [String, nil] the widget URL to the server that displays the amount of online members in a
+    # stylish way. `nil` if the widget is not enabled.
```
Comment (meew0) [inline lib/discordrb/data.rb]: Should also be indented (see below)

File: lib/discordrb/data.rb [hunk 46]
```diff
@@ -1468,6 +1597,35 @@ def voice_channels
...
+      return nil unless @embed
+      API.widget_url(@id)
+    end
+
+    # @param style_number [Integer] the style number of the widget picture  (For the diffrent styles just try which looks how)
+    # @return [String, nil] the widget banner URL to the server that displays the amount of online members,
+    # server icon and server name in a stylish way. `nil` if the widget is not enabled.
+    def widget_banner_url(style_number)
+      return nil unless @embed
+      API.widget_url(@id) + '?style=banner' + (style_number == Integer ? style_number.to_s : '1')
```
Comment (meew0) [inline lib/discordrb/data.rb]: This should be made part of the API method - everything dealing with the endpoint should be in there to achieve a clear separation.

Comment (meew0) [inline lib/discordrb/data.rb]: Also, the four types of banners aren't the only ones, there's also `shield`.

File: lib/discordrb/data.rb (hunk 42 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: you commented exactly the opposite yesterday...

File: lib/discordrb/data.rb (hunk 46 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: `shield` is the same as using `widget_url`

File: lib/discordrb/data.rb (hunk 42 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: https://github.com/meew0/discordrb/pull/181#discussion_r77004579

Comment (greenbigfrog) [inline lib/discordrb/data.rb]: Actually just my bad eyes...

Comment (greenbigfrog) [inline lib/discordrb/data.rb]: What should the full line look like? before I edit the wrong part again...

File: lib/discordrb/data.rb [hunk 47]
```diff
@@ -1468,6 +1597,35 @@ def voice_channels
...
     end
 
+    # @return [String, nil] the widget URL to the server that displays the amount of online members in a
+    # stylish way. `nil` if the widget is not enabled.
+    def widget_url
+      return nil unless @embed
+      API.widget_url(@id)
+    end
+
+    # @param style_number [Integer] the style number of the widget picture  (For the diffrent styles just try which looks how)
```
Comment (meew0) [inline lib/discordrb/data.rb]:
```
# @param style_number [Symbol] The style the picture should have. Possible styles are:
#
#   * `:banner1` creates a rectangular image with the server name, member count and icon, a "Powered by Discord" message on the bottom and an arrow on the right.
#   * `:banner2` creates a less tall rectangular image that has the same information as `banner1`, but the Discord logo on the right - together with the arrow and separated by a diagonal separator.
#   * `:banner3` creates an image similar in size to `banner1`, but it has the arrow in the bottom part, next to the Discord logo and with a "Chat now" text.
#   * `:banner4` creates a tall, almost square, image that prominently features the Discord logo at the top and has a "Join my server" in a pill-style button on the bottom. The information about the server is in the same format as the other three `banner` styles.
#   * `:shield` creates a very small, long rectangle, of the style you'd find at the top of GitHub `README.md` files. It features a small version of the Discord logo at the left and the member count at the right.
```

File: lib/discordrb/data.rb (hunk 46 above)
Comment (meew0) [inline lib/discordrb/data.rb]: What I would suggest is just changing the style_number parameter to a symbol and letting the user specify the style string themselves, like I indicated in the comment in my message above.

File: lib/discordrb/data.rb (hunk 47 above)
Comment (greenbigfrog) [inline lib/discordrb/data.rb]: Ty!

File: lib/discordrb/data.rb (hunk 42 above)
Comment (meew0) [inline lib/discordrb/data.rb]: `# @return [Array<Integration>] an array of all the integrations connected to this server.`

File: lib/discordrb/data.rb [hunk 48]
```diff
@@ -1535,11 +1707,11 @@ def update_voice_state(data)
         # Update the existing voice state (or the one we just created)
         channel = @channels_by_id[data['channel_id'].to_i]
         @voice_states[user_id].update(
-          channel,
-          data['mute'],
-          data['deaf'],
-          data['self_mute'],
-          data['self_deaf'])
+        channel,
```
Comment (meew0) [inline lib/discordrb/data.rb]: What is with this? This shouldn't be changed.

Comment (greenbigfrog) [inline lib/discordrb/data.rb]: stupid autoindent...
How do I find out which commit it was?

Comment (meew0) [inline lib/discordrb/data.rb]: `git blame`

Comment (greenbigfrog) [inline lib/discordrb/data.rb]: https://github.com/meew0/discordrb/pull/181/commits/3db3a2d114ac446ae17268cf76d110cb2ecbe7ef

**Comment by greenbigfrog** [comment]
Done again... Anything else too complain about?
